### ISSUE TITLE
bn.move_median seg fault

### ISSUE THREAD

**Issue by kwgoodman** [kwgoodman/bottleneck#111]
The following causes a seg fault:

```
In [1]: a = np.array([1, 2, np.inf, 3, 4, 5])
In [2]: bn.move_median(a, window=3)
Segmentation fault
```

If the location of np.inf or the window size is changed then the function does not seg fault.

move_median.c uses C's INFINITY to mark missing values in the heaps. I assume that is the problem. If I replace INFINITY with DBL_MAX (from float.h) and replace C's isinf with

```
inline int myisinf(double val)
{
    if (val == DBL_MAX || val == -DBL_MAX){return 1;}
    return 0;
}
```

then the seg fault disappears. However a new seg faults appears:

```
In [1]: m=np.finfo(np.float64).max
In [2]: a=np.array([m,m,m,m,m,m,m,m,m])
In [3]: bn.move_median(a, window=3)
Segmentation fault
```

So I traded one problem for another.

I'm not familiar with move_median.c. @jennolsen84 do you have any ideas on what to do?

**Comment by jennolsen84** [comment]
I just saw this, will take a look

**Comment by jennolsen84** [comment]
yes, you're right, I am converting nans to INF in the heap.

Some background: There are two heaps: one for values less than median, the other for values greater than the median.  a nan value can be stored as a -INF (if it is in the left heap) or +INF if it is in the right heap.  Also, as we rebalance the heaps, nan encoding can change from +INF to -INF, and vice versa.

So, the mapping is nan -> {+inf, -inf}.  I did not test with infinity inputs, and a bit more work would be needed.  I have a quick fix we can try in the short term, basically any input that is INF or -INF we can treat them as nans.  I am not 100% sure, but it seems like the output from pandas for the array given is:

```
[ nan  nan  nan  nan  nan   4.]
```

If we change https://github.com/kwgoodman/bottleneck/blob/014863934dfb1af4388d8ae6ec24028c33e26a99/bottleneck/src/auto_pyx/csrc/move_median.c#L617 to:

```
int is_nan_val = isnan(val) | isinf(val);
```

and https://github.com/kwgoodman/bottleneck/blob/014863934dfb1af4388d8ae6ec24028c33e26a99/bottleneck/src/auto_pyx/csrc/move_median.c#L768 to:

```
if (isnan(val) | isinf(val))
```

it works on the given sample input.  I haven't tried to do an exhaustive test yet (where I just randomly generate matrices and compare the output vs pandas).

note: I used a | bitwise operator, instead of logical operator || above.  This way, the hope is that the hardware will evaluate isnan() and isinf() in parallel and just do a bitwise or.

Unfortunately, I am traveling for the next few days, and will have some time May 12ish onwards.  I will be able to look at it a bit more then.

Hope this helps though.

**Comment by kwgoodman** [comment]
I've been refactoring move_median.c (the move_median_refactor branch). I figured refactoring would be a gentle introduction to the code. Once I'm up to speed on the code (which will take quite a while) I'll re-read your comment (thank you for all the details) and ask you lots of questions!

Without yet understanding the code I was wondering if the node could contain a is_nan variable to mark missing values. Then we could distinguish between NaN and Inf. If possible I'd like to match the behaviour of bn.slow.move_median.

**Comment by jennolsen84** [comment]
yes, that should work.  I would probably call the variable something else, so you can differentiate between nan, +INF and -INF.  And you'll need to modify get_median function too.  So that if the array is all INF or -INF, the correct answer gets reported.  I would also look at the size of the new variable [1 byte should be enough] and place it in the struct definition appropriately so that struct is packed nicely (and there is no un-necessary padding inserted by the compiler).

Other impacts might be that most of the time that value will not be used, so your working set in memory will be higher, and possibly more cache misses will occur, thus a slight degradation in performance.

If you have any questions about the code, let me know and I will try to answer quickly

**Comment by kwgoodman** [comment]
bn.move_median can now handle INF in the input array. It also runs faster.

I ended up doing a complete rewrite of bn.move_median. Instead of placing NaNs in the heaps (after changing NaN to +/-INF) I put them in a nan array. Thus (1) there was no need to sort the NaNs and (2) the heaps became smaller in the presence of NaNs.

In the bottleneck benchmark with a 1000 by 1000 array with 1/3 NaNs, the new approach was more than twice as fast.

**Comment by jennolsen84** [comment]
Impressive.  What was your window size (performance can vary with it)?  How did you end up testing it? 

I ended up just running it in a loop with randomized data and testing it vs pandas.  That shook out a lot of the bugs.

Also, will a median of `[-INF, INF, INF]` be INF or NAN?

**Comment by kwgoodman** [comment]
There are two versions of every bottleneck function: bn.func() and bn.slow.func(). So I tested bn.move_median() against bn.slow.move_median(). I added this unit test to bottleneck:

```
# ---------------------------------------------------------------------------
# move_median.c is complicated. Let's do some more testing.
#
# If you make changes to move_median.c then do lots of tests by increasing
# range(100) in the two functions below to range(10000).

def test_move_median_with_nans():
    "test move_median.c with nans"
    fmt = '\nfunc %s | window %d | min_count %s\n\nInput array:\n%s\n'
    aaae = assert_array_almost_equal
    min_count = 1
    size = 8
    func = bn.move_median
    func0 = bn.slow.move_median
    rs = np.random.RandomState([1, 2, 3])
    for i in range(100):
        a = np.arange(size, dtype=np.float64)
        idx = rs.rand(*a.shape) < 0.1
        a[idx] = np.inf
        idx = rs.rand(*a.shape) < 0.2
        a[idx] = np.nan
        rs.shuffle(a)
        for window in range(2, size + 1):
            actual = func(a, window=window, min_count=min_count)
            desired = func0(a, window=window, min_count=min_count)
            err_msg = fmt % (func.__name__, window, min_count, a)
            aaae(actual, desired, decimal=5, err_msg=err_msg)
```

When the unit test failed I'd copy the failing input array to move_median_debug.c which sets the number of children to 2 and prints the heaps at each step in binary tree form so that I could see what was going on. I did the debugging in pure C (i.e. no cython) using the new makefile (gdb and valgrind).

I found that optimizing the number of children depended on window size.

Because you asked I wrote a script to measure performance:

```
import numpy as np
import bottleneck as bn
from bottleneck.benchmark.autotimeit import autotimeit

rs = np.random.RandomState([1, 2, 3])
a = rs.rand(1000000)
a[a < 0.2] = np.nan
for exponent in range(1, 6):
    window = 10**exponent
    t = autotimeit(lambda: bn.move_median(a, window=window, axis=-1))
    print "%6d %f" % (window, t)
```

And I got a seg fault! I'm glad you asked. Back to debugging. I've been hand coding tiny test arrays. Not sure how to deal with such a large array in pure C. 

Oh, BTW:
```
In [1]: bn.move_median([-np.inf, np.inf, np.inf], window=3, min_count=1)
Out[1]: array([-inf,  nan,  inf])
```

**Comment by kwgoodman** [comment]
I fixed the seg fault. Here's the performance versus window size for the code in master (using the above script on my slow laptop).

With no NaNs:
```
window  seconds
    10  0.042365
   100  0.068565
  1000  0.082538
 10000  0.107702
100000  0.115240
```

20% nans:
```
window  seconds
    10  0.052971
   100  0.075063
  1000  0.089599
 10000  0.105983
100000  0.128464
```
All nans:
```
window  seconds
    10  0.008539
   100  0.008545
  1000  0.008521
 10000  0.008538
100000  0.009643
```

Here are the corresponding results for PR #110:

```
window  seconds
    10  0.081875
   100  0.112471
  1000  0.130629
 10000  0.158817
100000  0.169358

window  seconds
    10  0.088468
   100  0.120596
  1000  0.141156
 10000  0.169303
100000  0.185353

window  seconds
    10  0.049541
   100  0.048705
  1000  0.048004
 10000  0.048061
100000  0.051045
```

Before completely rewriting move_median.c for NaN input, I did some optimization of PR #110:

```
window  seconds
    10  0.054325
   100  0.080486
  1000  0.094239
 10000  0.118666
100000  0.129393

window  seconds
    10  0.060485
   100  0.088697
  1000  0.104200
 10000  0.127559
100000  0.131378

window  seconds
    10  0.028171
   100  0.027808
  1000  0.027095
 10000  0.027278
100000  0.029650
```

**Comment by jennolsen84** [comment]
👍  looking forward to the release

**Comment by kwgoodman** [comment]
How big are the arrays you use? I've been toying with some convenience functions for threading. For example, here's the timing without threads:

```
In [1]: a = np.random.rand(4000, 4000)
In [2]: a[a < 0.3] = np.nan
In [3]: timeit bn.move_median(a, window=500, min_count=1)
1 loops, best of 3: 1.37 s per loop
```

Timing with four threads:

```
In [4]: from bottleneck.threads import p_move_median, p_nansum, start_pool
In [5]: start_pool(4)
In [6]: timeit p_move_median(a, window=500, min_count=1)
1 loops, best of 3: 596 ms per loop
```

The input arrays should be large and/or the function should be slow to make up for the overhead of dealing with the threads. The function p_move_median automatically chunks the data into nthread pieces.

**Comment by jennolsen84** [comment]
```o_O``` bottleneck.threads? awesome!

I don't remember the input sizes, but I did anywhere from size 1 to a big one where move_median took several seconds to compute.

Threading makes a lot of sense, here are a few thoughts:

In case of 4000x4000, it is easy to split the data on the axis, good win there.  But what about the other extreme? 1D case or 2D case with 2 rows but large number of columns (using 8 threads).  Then there is some overlap in the computation, but it's an interesting problem.

Another (slightly more advanced case) would be if somehow the chunking was exposed, so that I would easily be able to many of these algorithms with my own thread pool.  This is useful when there is a bunch of other work (possibly non bottleneck related) that can be scheduled and things work cooperatively w/o oversubscribing the cores.

**Comment by kwgoodman** [comment]
Great idea! It didn't yet occur to me to chunk along more than one axis. I did try a (2, 3, 4) input array and the code only made two chunks along axis 0 even though there were 4 threads.

There's also the 1d input array case of moving window functions. At the moment the code runs those in a single thread. Chunking those is more complicated because the chunks would have to overlap by the width of the window. If the window is a sizeable fraction of the length of the array then it is probably better to use one chunk unless you have lots of cores and nothing else to run in the meantime.

Some functions, like bn.median, can't be chunked for the 1d input array case because the median of several sub medians is not the median. But if you are finding the median of millions of numbers then the difference doesn't matter and we could provide an override.

The functions that I have written so far are convenience functions. They block until all threads are done. As soon as one thread is done we should be moving on to the next job. I think it is best to start with easy to use functions to get people to try them and to figure out how things work behind the scenes.

If we exposed a chunker function, we'd also need a dechunker. It is interesting to think how that would work.

I'm working on all this in the threads branch.
